### PR TITLE
PP-2784 add relationship cards3ds to transactions 2

### PR DESCRIPTION
**Pull request by chrisgrimble**
Currently the card_3ds table references the charge_id. As we will be
deleting the charges tables we want to reference the transaction_id of
the ChargeTransaction.

- Added a link back to chargeTransaction for Card3ds data
- Back fill script for transaction_id on card_3ds

### CODE REVIEW

File: src/main/java/uk/gov/pay/connector/service/CardAuthoriseService.java [hunk 1]
```diff
@@ -22,6 +22,7 @@
 import uk.gov.pay.connector.model.domain.ChargeStatus;
 import uk.gov.pay.connector.model.domain.GatewayAccountEntity;
 import uk.gov.pay.connector.model.domain.PaymentRequestEntity;
+import uk.gov.pay.connector.model.domain.transaction.ChargeTransactionEntity;
 import uk.gov.pay.connector.model.gateway.AuthorisationGatewayRequest;
 import uk.gov.pay.connector.model.gateway.GatewayResponse;
```
Comment (georgeracu) [inline src/main/java/uk/gov/pay/connector/service/CardAuthoriseService.java:27]: On line 47 there is unused field `private final Card3dsDao card3dsDao`. Needs removed

File: src/main/java/uk/gov/pay/connector/tasks/MigrateAddTransactionIdToCard3dsTask.java [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+    @Inject
+    public MigrateAddTransactionIdToCard3dsTask(AddTransactionIdToCard3dsWorker worker) {
+        super(TASK_NAME);
+        this.worker = worker;
+    }
+
+    @Override
+    public void execute(ImmutableMultimap<String, String> parameters, PrintWriter output) {
+        worker.execute();
```
Comment (georgeracu) [inline src/main/java/uk/gov/pay/connector/tasks/MigrateAddTransactionIdToCard3dsTask.java]: Should have the id where to start from

File: src/test/java/uk/gov/pay/connector/it/dao/Card3dsDaoITest.java [hunk 3]
```diff
@@ -46,6 +52,9 @@ public void shouldCreateACardEntity() throws Exception {
                 .build();
 
         chargeDao.persist(chargeEntity);
+        PaymentRequestEntity paymentRequest = PaymentRequestEntity.from(chargeEntity, ChargeTransactionEntity.from(chargeEntity));
+        paymentRequestDao.persist(paymentRequest);
+
 
         Card3dsEntity card3dsEntity = Card3dsEntity.from(chargeEntity);
```
Comment (georgeracu) [inline src/test/java/uk/gov/pay/connector/it/dao/Card3dsDaoITest.java:60]: Shouldn't this have a test if a card3ds belongs to a charge transaction?

File: src/test/java/uk/gov/pay/connector/it/dao/DatabaseFixtures.java [hunk 4]
```diff
@@ -508,7 +508,7 @@ public String getSecureRedirectToken() {
 
     public class TestRefund {
         Long id = RandomUtils.nextLong(1, 99999);
-        String externalRefundId = RandomIdGenerator.newId();
+        String externalRefundId = RandomIdGenerator.newId().trim();
```
Comment (georgeracu) [inline src/test/java/uk/gov/pay/connector/it/dao/DatabaseFixtures.java]: Thus RandomIdGenerator should be fixed in a separate PR as it generates ids with less characters and the db will add spaces

**Review by georgeracu — CHANGES_REQUESTED** [review]
A few small changes required

File: src/main/java/uk/gov/pay/connector/service/CardAuthoriseService.java (hunk 1 above)
Comment (chrisgrimble) [reply src/main/java/uk/gov/pay/connector/service/CardAuthoriseService.java:27]: Done

File: src/test/java/uk/gov/pay/connector/it/dao/DatabaseFixtures.java (hunk 4 above)
Comment (chrisgrimble) [reply src/test/java/uk/gov/pay/connector/it/dao/DatabaseFixtures.java]: Done

File: src/main/java/uk/gov/pay/connector/tasks/MigrateAddTransactionIdToCard3dsTask.java (hunk 2 above)
Comment (chrisgrimble) [reply src/main/java/uk/gov/pay/connector/tasks/MigrateAddTransactionIdToCard3dsTask.java]: Done

File: src/test/java/uk/gov/pay/connector/it/dao/Card3dsDaoITest.java (hunk 3 above)
Comment (chrisgrimble) [reply src/test/java/uk/gov/pay/connector/it/dao/Card3dsDaoITest.java:60]: Not needed as discuss with George

**Review by georgeracu — APPROVED** [review]
